src/wayland/shell/mod.rs

```diff
@@ -14,5 +14,29 @@
 //! - The [`legacy`](legacy/index.html) module provides handlers for the `wl_shell` protocol, which
 //!   is now deprecated. You only need it if you want to support apps predating `xdg_shell`.
 
+use super::Serial;
+use thiserror::Error;
+
 pub mod legacy;
 pub mod xdg;
+
+/// Represents the possible errors returned from
+/// a surface operation
+#[derive(Debug, Error)]
+pub enum SurfaceError {
+    /// The underlying `WlSurface` is no longer alive
+    #[error("the surface is not alive")]
+    SurfaceNotAlive,
+}
+
+/// Represents the possible errors returned from
+/// a surface ping
+#[derive(Debug, Error)]
+pub enum PingError {
+    /// The operation failed because the underlying surface has an error
+    #[error("the ping failed cause the underlying surface has an error")]
+    SurfaceError(#[from] SurfaceError),
```

Comment:
Given the underlying error has only 1 variant (and is not expected to gain any), this really should just be a `DeadSurface` error, without any nesting.

Reply:
Removed the nesting, now it is just a DeadResource variant.